neo/Consensus/ConsensusService.cs

```diff
@@ -336,37 +306,82 @@ private void OnPrepareResponseReceived(ConsensusPayload payload, PrepareResponse
             CheckSignatures();
         }
 
-        private void OnTimeout(object state)
+        protected override void OnReceive(object message)
         {
-            lock (context)
+            switch (message)
             {
-                if (timer_height != context.BlockIndex || timer_view != context.ViewNumber) return;
-                Log($"timeout: height={timer_height} view={timer_view} state={context.State}");
-                if (context.State.HasFlag(ConsensusState.Primary) && !context.State.HasFlag(ConsensusState.RequestSent))
-                {
-                    Log($"send perpare request: height={timer_height} view={timer_view}");
-                    context.State |= ConsensusState.RequestSent;
-                    if (!context.State.HasFlag(ConsensusState.SignatureSent))
-                    {
-                        context.Timestamp = Math.Max(DateTime.Now.ToTimestamp(), Blockchain.Default.GetHeader(context.PrevHash).Timestamp + 1);
-                        context.Signatures[context.MyIndex] = context.MakeHeader().Sign(context.KeyPair);
-                    }
-                    SignAndRelay(context.MakePrepareRequest());
-                    timer.Change(TimeSpan.FromSeconds(Blockchain.SecondsPerBlock << (timer_view + 1)), Timeout.InfiniteTimeSpan);
-                }
-                else if ((context.State.HasFlag(ConsensusState.Primary) && context.State.HasFlag(ConsensusState.RequestSent)) || context.State.HasFlag(ConsensusState.Backup))
+                case Start _:
+                    OnStart();
+                    break;
+                case Timer timer:
+                    OnTimer(timer);
+                    break;
+                case Blockchain.PersistCompleted completed:
+                    OnPersistCompleted(completed.Block);
+                    break;
+                case LocalNode.InventoryReceived received:
+                    OnInventoryReceived(received.Inventory);
+                    break;
+            }
+        }
+
+        private void OnStart()
+        {
+            Log("OnStart");
+            LocalNode.Singleton.Blockchain.Tell(new Blockchain.Register());
+            Context.Parent.Tell(new LocalNode.Register());
+            InitializeConsensus(0);
+        }
+
+        private void OnTimer(Timer timer)
+        {
+            if (timer.Height != context.BlockIndex || timer.ViewNumber != context.ViewNumber) return;
+            Log($"timeout: height={timer.Height} view={timer.ViewNumber} state={context.State}");
+            if (context.State.HasFlag(ConsensusState.Primary) && !context.State.HasFlag(ConsensusState.RequestSent))
+            {
+                Log($"send prepare request: height={timer.Height} view={timer.ViewNumber}");
+                context.State |= ConsensusState.RequestSent;
+                if (!context.State.HasFlag(ConsensusState.SignatureSent))
                 {
-                    RequestChangeView();
```

Comment:
Where and when does Primary/Speaker change view now?

Reply:
https://github.com/neo-project/neo/blob/8d0633a9f081e2f7ec19d4bf9c00e253fc22fcaa/neo/Consensus/ConsensusService.cs#L352-L355